src/Clash/Cores/Rgmii.hs

```diff
@@ -0,0 +1,216 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+
+{- |
+Module      : Clash.Cores.Rgmii
+Description : Functions and types to connect an RGMII PHY to a packet stream interface.
+
+To keep this module generic, users will have to provide their own "primitive" functions:
+
+    1. delay functions to set to the proper amount of delay (which can be different for RX and TX);
+    2. iddr function to turn a single DDR (Double Data Rate) signal into 2 non-DDR signals;
+    3. oddr function to turn two non-DDR signals into a single DDR signal.
+
+Note that Clash models a DDR signal as being twice as fast, thus both facilitating
+and requiring type-level separation between the two "clock domains".
+-}
+module Clash.Cores.Rgmii (
+  RgmiiRxChannel (..),
+  RgmiiTxChannel (..),
+  rgmiiReceiver,
+  rgmiiSender,
+  unsafeRgmiiRxC,
+  rgmiiTxC,
+) where
+
+import Clash.Prelude
+
+import Protocols
+import Protocols.PacketStream
+
+import Data.Maybe (isJust)
+
+-- | RX channel from the RGMII PHY
+data RgmiiRxChannel dom domDDR = RgmiiRxChannel
+  { rgmiiRxClk :: "rx_clk" ::: Clock dom
+  , rgmiiRxCtl :: "rx_ctl" ::: Signal domDDR Bit
+  , rgmiiRxData :: "rx_data" ::: Signal domDDR (BitVector 4)
+  }
+
+instance Protocol (RgmiiRxChannel dom domDDR) where
+  type Fwd (RgmiiRxChannel dom domDDR) = RgmiiRxChannel dom domDDR
+  type Bwd (RgmiiRxChannel dom domDDR) = Signal dom ()
+
+-- | TX channel to the RGMII PHY
+data RgmiiTxChannel domDDR = RgmiiTxChannel
+  { rgmiiTxClk :: "tx_clk" ::: Signal domDDR Bit
+  , rgmiiTxCtl :: "tx_ctl" ::: Signal domDDR Bit
+  , rgmiiTxData :: "tx_data" ::: Signal domDDR (BitVector 4)
+  }
+
+instance Protocol (RgmiiTxChannel domDDR) where
+  type Fwd (RgmiiTxChannel domDDR) = RgmiiTxChannel domDDR
+  type Bwd (RgmiiTxChannel domDDR) = Signal domDDR ()
```

Comment:
Aren't you basically abusing the ODDR primitive here as a clock generator? Maybe we should add a wrapper for the ODDR primitive that outputs a `Clock dom` rather than a `Signal domDDR Bit` (note difference in domain, factor two period difference).

I don't think there's anything wrong with using an ODDR primitive to generate a clock. But Clash is strongly typed and while there is a certain isomorphism, I still feel this outputs a `Clock`, not a `Signal`. It is _used_ as a time reference rather than being relative to one. The `Signal domDDR`s have a setup and hold constraint, the clock output does not. And more arguments revolving around the same principle.

Reply:
We could add a function
```haskell
oddrClockGen ::
  forall dom domDdr .
  (DomainPeriod dom ~ (*) 2 (DomainPeriod domDdr)) =>
  Clock dom ->
  Reset dom ->  
  ( Clock dom ->
    Reset dom ->
    Enable dom ->
    Bit ->
    Signal dom Bit ->
    Signal domDdr Bit
  ) ->
  Clock dom
```
where we construct the function such that in HDL, it outputs the output of the DDR register while in Haskell it just outputs a `Clock`.

Reply:
> From the FPGA view it's actually no longer clock and can't be used as such.

I don't think you can use the output of the ODDR primitive, period. You can only wire it to a pin, not to any internal logic. So that remains the same whether it's a `Signal` or a `Clock`: it cannot be used at all.

Reply:
True perhaps it's better to make it neither a clock or a signal. But rather something like `WorldSignal` or something. Which is just a newtype around signal which you can only route.

Reply:
What's the status on this? I like the idea of having a Signal type that can only be routed. Changing it to a clock doesn't make sense to me.

Reply:
The status from my point of view is that I was waiting for consensus or some form of decision of what to do.

Why doesn't the clock make sense to you? To me, the signal doesn't really make sense for the reasons I already indicated. The output of the DDR primitive is used as a clock for external circuitry and that is its only purpose.

Reply:
I'd say just outputting `Clock dom` is not qualitatively different than outputting the ODDR-generated clock; it's just a different construction inside the FPGA that makes stuff easier. Even Xilinx [recommends it here](https://docs.amd.com/r/en-US/ug949-vivado-design-methodology/Creating-an-Output-Clock); one of the first search results I found.